### DISCUSSION TITLE
Will Red Knot be a drop-in replacement for mypy or pyright?

### DISCUSSION THREAD

**Discussion by kkpattern** [astral-sh/ruff#15149] [Q&A] [answered]
Hi, thanks to all of you for the amazing work on Ruff and uv.

One reason we were able to quickly adopt Ruff and uv is that they are drop-in replacements for previous tools like black and pip.

However, the world of type checker tools seems to be a bit more complex. Currently, mypy and pyright are among the most widely used type checkers, but they have different philosophies. For example, our team depends on mypy's setting "--disallow-untyped-defs" to enforce type hints, while some other teams in our company prefer pyright for its superior type inference and speed.

I'm curious about which tool Red Knot is more similar to, and maybe we can prepare for the migration in advance. Regardless, I believe Red Knot will be an amazing type checker, and I thank you in advance.

Happy holidays!

**Comment by carljm** [comment, marked as answer]
Type checkers are complex tools, and red knot will certainly not be a drop-in replacement for either mypy or pyright. It will be a distinct alternative with its own design choices. We prioritize good type inference, even in un-annotated code, and we aim to avoid false positives on working unannotated code, where mypy is often more willing to say “you must write this (working) code differently.”

We will want to offer something like `disallow-untyped-defs`, but we would consider that a lint rule to be built on top of the core red knot engine in a type-aware linter, not a core type checker feature.

**Comment by danlamanna** [comment]
One philosophical decision that's important for Django users are plugin systems. mypy supports plugins which help significantly in taming Django's dynamism (one of the places where type checking can help the most). Pyright on the other hand, does not[^1].

I think I know the answer, but has red knot committed to a particular viewpoint on this?

[^1]: https://microsoft.github.io/pyright/#/mypy-comparison?id=plugins

**Reply by carljm** [reply]
We have not committed to a viewpoint on this, but we are also not prioritizing a plugin system in the short or medium term.

(In general, I am sympathetic to the viewpoint that plugins are antithetical to a well-specified and interoperable type system, and it is preferable -- though higher-investment -- to improve the type system such that it can express important or common dynamic patterns.)

**Reply by kkom** [reply]
> (In general, I am sympathetic to the viewpoint that plugins are antithetical to a well-specified and interoperable type system, and it is preferable -- though higher-investment -- to improve the type system such that it can express important or common dynamic patterns.)

This is fantastic to hear! I'm very happy with and grateful for `uv`'s deep support for latest Python standards, and I would love it if the type checker followed a similar philosophy! Thank you @carljm!
